Defer bug report processing

alexfu opened the pull request:
Resolves #18 and #4. The new bug report processing functionality will use `JobScheduler` on devices running Lollipop or above. Everything below will fallback to using a standard `IntentService` (`SubmitReportLegacyService`).

For Lollipop and above devices, `JobScheduler` will reschedule jobs that fail due to a network loss. When using the `SubmitReportLegacyService`, if a bug report submission fails due to a network loss, it will be retried upon the next Buglife initialization (usually when the app opens again).

alexfu committed:
Add skeleton structure for implementing JobScheduler.

alexfu committed:
Implement JobScheduler and introduce new BugReporter construct.

alexfu committed:
Decouple JobScheduler from SubmitReportTask.

alexfu committed:
Update parcelable implementation.

alexfu committed:
Add missing service declaration.

alexfu committed:
Rearrange and clean up a bit.

alexfu committed:
Implement a reporting service for legacy reporting.

Legacy reporting will attempt to submit the report, however if it fails it will attempt again on the next SDK initialization.

alexfu committed:
Remove obsolete RequestHandler and obsolete pending reports code.

alexfu committed:
Add licensing info to new files.

alexfu committed:
Refactor and clean up.

schukin commented on src/main/java/com/buglife/sdk/BugReporterImpl.java:
```diff
@@ -0,0 +1,87 @@
+            String jsonReport = report.toJSON().toString();
+            SubmitReportLegacyService.start(mContext, jsonReport);
+        } catch (JSONException e) {
+            Log.e("Failed to encode report!", e);
```
Can we bubble up this error somehow?

alexfu replied:
@schukin Do you mean surfacing this error to the UI? We could display a Toast or re-throw the exception. My initial thought would be to re-throw since the having a badly formatted report would be a pretty serious error.

schukin replied:
Yeah, we should surface this error to the UI. That way when users encounter it, they at least have a vague clue that something went wrong and can help us figure it out. I would think it'd make sense to re-throw the exception at this level. Up the chain, I think we'd either want to throw a RuntimeException & let the app crash, or show a Toast so the user knows something went wrong. Thoughts?

alexfu replied:
@schukin would you be opposed to doing both, throwing the exception and showing a toast here?

schukin replied:
@alexfu that sounds reasonable 👍

schukin commented on src/main/java/com/buglife/sdk/reporting/SubmitReportLegacyService.java:
```diff
@@ -0,0 +1,175 @@
+    @Override protected void onHandleIntent(@Nullable Intent intent) {
+        File cacheFile = getReportsCacheFile(getApplicationContext());
+        if (!cacheFile.exists()) {
+            Log.i("Reports cache file doesn't exist! No reports to submit.");
```
Whenever I try to submit a report on my KitKat device, I see this get logged & my bug report is never actually submitted. It seems the `reports` directory for legacy submission is never actually created, so this method returns early. Before I went ahead and changed anything, I wanted to verify that the intent here (no pun intended) is to submit the "current" bug report, as well as all cached bug reports?

alexfu replied:
@schukin This is a good catch. I had changed the behavior slightly so this check is no longer valid.

schukin reviewed (changes requested):
This is *awesome*. Love seeing `BugContext` split into `SessionSnapshot` & `DeviceSnapshot` too 👌 Just two questions before I merge.

alexfu committed:
Fix cache file check.

Since reports are not persisted until they fail, this check is no longer
valid.

alexfu committed:
Re-throw JSON exception and display error.

alexfu commented:
@schukin updated to address changes.

alexfu committed:
Don't throw exception since it will prevent Toast from showing.

schukin commented:
w00t 🤘
